File: Practice/src/stack_queue/BridgeTruck.js
// 다리를 지나는 트럭
function solution(bridge_length, weight, truck_weights) {
	let answer = 0;
	let bridge=new Array(bridge_length).fill(0);
	while(truck_weights.length>0){
		let sum=bridge.reduce((a,b)=>a+b);
		if(weight-truck_weights[0]>=sum-bridge[0]) {
			sum+=truck_weights[0];
			bridge.push(truck_weights.shift());
			bridge.shift();
		} else if(bridge[0]>0){
			bridge.push(0);
			bridge.shift();
		} else{
			let find=bridge.find(e=>e>0);
			let index=bridge.findIndex(e=>e>0);
			bridge=Array.prototype.concat(bridge.slice(index+1), new Array(index).fill(0));
			if(sum+truck_weights[0]-find<=weight){
				bridge.push(truck_weights.shift());
			}else {
				bridge.push(0);
			}
			answer+=index;
		}
		if(sum==0)continue;
		answer++;
	}
	return answer+bridge_length;
}

File: Practice/src/stack_queue/BridgeTruck.java
package stack_queue;

import java.util.Stack;

public class BridgeTruck {
	//28.6 / 100.0
    public int solution(int bridge_length, int weight, int[] truck_weights) {
		int answer=1;
		int start=0;
		int end=0;
		int load=0;
		Stack<Integer> s=new Stack<Integer>();
		for(int i=0;i<truck_weights.length;i++) {
			s.add(0);
		}
		while(s.lastElement()!=bridge_length) {
			answer++;
			load=0;
			int i=start;
			do {
				load+=truck_weights[i];
				s.set(i, s.get(i)+1);
				i++;
			}while(i<=end);
			if(i<s.size()&&weight>=load+truck_weights[i]) end++;
			if(s.get(start)>=bridge_length) {
				if(start==end) end++;
				start++;
			}
		}
        return answer;
    }
}
